# K-Nearest Neighbour Implementation

In [1]:
#Importing required library
import pandas as pd
import numpy as np
from collections import Counter


## Dataset

For this tutortial we will use the Iris dataset which is opensource and can be found at the following link :

https://www.kaggle.com/uciml/iris


In [2]:
#Setting names for the csv header
headernames = ['sepal-length', 'sepal-width', 'petal-length', 'petal-width', 'Class']

In [3]:
#opening the csv file
dataset = pd.read_csv("iris.csv", names = headernames)
dataset.head()

sepal-length  sepal-width  petal-length  petal-width        Class
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

Complete details of the dataset can be read on it's source

In [4]:
#Seperating teh input features and outout labels
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, 4].values

In [5]:
#converting text labels to numeric form
labels, unique = pd.factorize(y)

In [6]:
#splitting data in test and train segments
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, labels, test_size = 0.40)

In [7]:
y_test

array([2, 2, 0, 2, 1, 1, 1, 1, 0, 2, 2, 0, 2, 1, 2, 0, 0, 2, 1, 2, 1, 0,
       2, 1, 1, 0, 0, 2, 1, 2, 0, 1, 1, 1, 0, 0, 0, 2, 0, 0, 2, 1, 0, 2,
       2, 1, 2, 0, 1, 1, 1, 0, 0, 0, 2, 0, 2, 1, 0, 0], dtype=int64)

# Understanding the Algoriothm

The K-Nearest Neighbour Algorithm (KNN for short) is a supervised machine learning algorithm.

Being **supervised** means that it requires training labels to learn the dataset.

The algorithm is suprisingly simple and short. 

Imagine you have 3 sacks, each filled with balls of a different shade of <font color='green'>green</font>. You also have a sack with balls of all 3 shades of <font color='green'>green</font> mixed together. Your task is to place those mixed balls in their respective sack. 
The simplest approach would be to pick out each ball and compare it against balls of each sack and place it where it color matches the most balls. 
What you have done here is quite similar to the KNN algorithm.

***To understand the algorithm in terms of computations you need to know the concept of euclidean distance(Not covered in this tutorial).***

We have a set of training data points which have labels against them, and a set of test data points(which need classification). We take a data point from the test set and calculate its euclidean distance with every point in the training set. We then pick the top N smallest distances and see against which classes our test point has had the smallest distance. The class which appears most is class we will associate with our test data point. 

Quite simple right?

Lets get to coding then.

In [8]:
#Creating a helper function
def takeSecond(elem):
    return elem[1]

No need to worry about the function above, it is just a helper function we will use in the algorithm to help with sorting.
Now let's write the algorithm.

In [11]:
def KNNClassify(X_test, Y_train = y_train,X_train = X_train, k = 8):
   
    min_dist = []
    #for every example in the training set, calculate eucledien distance against the test example
    for i,point in enumerate(X_train):
        d1 = (point[0]-X_test[0])**2
        d2 = (point[1]-X_test[1])**2
        d3 = (point[2]-X_test[2])**2
        d4 = (point[3]-X_test[3])**2
        dist = np.sqrt(d1+d2+d3+d4)
        #append the calculated distance in a list
        min_dist.append((i,dist))
    #sort distances in ascending order    
    min_dist.sort(key = takeSecond)
    
    #get top k nearest neighbours
    neighbours = min_dist[:k]
    #get index of the minimum distances
    idx = []
    for tup in neighbours:
        idx.append(tup[0])
    #check which label has majority
    output = Y_train[idx]
    values, counts = np.unique(output, return_counts=True)
    #return label with majority occurence
    max_idx = np.argmax(counts)
    return values[max_idx]

**ALL DONE!!**
To test it allw e need to do is send our test data points one by one. Good thing we have a python built-in function called *map* to send multiuple inputs to a fuunction. Let's use it.

In [12]:
predictions = list(map(KNNClassify, X_test))

Now lets see how accurate of a prediction our model made.

In [13]:
def accuracy(pred , y_test):
    count = 0
    for i in range(len(pred)):
        if pred[i] == y_test[i]:
            count +=1
            
    return print("Accuracy =", (count/len(pred))*100, "%")

In [14]:
accuracy(predictions, y_test)

Accuracy = 96.66666666666667 %


**96.67%** is a pretty good number if you ask me.